### ISSUE TITLE
Save related models where foreign key is not nullable

### ISSUE THREAD

**Issue by thijsvdanker** [christophrumpel/laravel-factories-reloaded#9]
The current implementation of `with()` and the examples in the docs only work when the foreign key is `nullable()` like *group_id* on the recipes table.

```php
        Schema::create('recipes', function (Blueprint $table) {
            $table->bigIncrements('id');
            $table->unsignedBigInteger('group_id')->nullable();
```

When it is not nullable it is not possible to create the related model before using it in `saveMany()`.

A function like `make()` on Laravel's [FactoryBuilder](https://github.com/laravel/framework/blob/6.x/src/Illuminate/Database/Eloquent/FactoryBuilder.php#L229) would probably help.

Would this be the right direction or is there another clean way you can think of?

**Comment by christophrumpel** [comment]
hey, good catch! We definitely need the `make` method as well. I guess this is the best way to solve this problem. Is this something you would want to work it? If not I will start today or tomorrow. I also need that `make` method for other stuff.

**Comment by thijsvdanker** [comment]
I will certainly not get to this until next week so if you could work on it that would be 💯!

**Comment by christophrumpel** [comment]
👍

**Comment by christophrumpel** [comment]
Creating the `make` method was the easy part. I also need to change the `times` method now because it needs to work for both `create` and `make`, but I cannot change the `create` method, because I love that I get a model instance back here and my IDE knows it. I could add `createMultiple` and `makeMultiple` methods but then they need two params for count and array to override which does not look nice.. 🤔

**Comment by christophrumpel** [comment]
Came up with a new solution. `times` will now give back a new object where you need to chain `create`. Also the `make` method is now included: https://github.com/christophrumpel/laravel-factories-reloaded/pull/10 which means the issue should be gone now too.

**Comment by stephanus-tantiono** [comment]
hai, can you give example how to save related models where foreign key is not nullable ?,
maybe you can also show RecipeFactory too in doc for easier understanding

**Comment by christophrumpel** [comment]
Hey @stephanus-stuff,

have you checked this section of the docs already?
https://github.com/christophrumpel/laravel-factories-reloaded#relations

When you want to save it like this, use the `make` method for the related model before you add it:

```php
$user = UserFactory::new()->create();
$user->recipes()->saveMany(RecipeFactory::new()->times(4)->make());
```

if you use the baked in `with` method of the package you just do it like this:

```php
$user = UserFactory::new()
    ->with(Recipe::class, 'recipes')
    ->create();
```
(You just need to make sure the RecipeFactory class also already exists, created through this package)

Did this solve your problem?

**Comment by stephanus-tantiono** [comment]
Hai @christophrumpel thanks for the answer, 
to make it work, RecipeFactory need to set the getData isn't it ? CMIIW I'm still learning

````
    public function getData(Generator $faker): array
    {
        return [
            'user_id' => 'I WANT EXAMPLE THIS PART',
             ...
        ];
    }
````

in normal factory usually I use
````
factory(User::class)->create()
````

**Comment by christophrumpel** [comment]
so, I'm not sure I know exactly what you mean. What works in your application with `factory(User::class)->create()`, works with my package like `UserFactory::new()->create()`. So these two are the same.

Ah I might know what you mean. In `getData` you need to define default data, like you did in or Laravel factory file.

**Comment by stephanus-tantiono** [comment]
> 
> 
> so, I'm not sure I know exactly what you mean. What works in your application with `factory(User::class)->create()`, works with my package like `UserFactory::new()->create()`. So these two are the same.
> 
> Ah I might know what you mean. In `getData` you need to define default data, like you did in or Laravel factory file.

yeah, then what I should define for foreign key(user_id) in getData at RecipeFactory ?

**Comment by christophrumpel** [comment]
What do you define there when you use laravel factories? This is up to you. You can just put `1` this would work. You could also leave it blank, this is what I usually do. if this model has a foreign key, it belongs to a different model. So let's say we are talking about a recipe here and it has a `user_id` field. Then in my test I would first `create` a new user. (does not have a foreign key) and then I would create recipe with the id of the user.

```php
$user = UserFactory::new()->create(); // you can use create because no foreign key on this model

// Option 1: Then you can create a new recipe and add the user id
RecipeFactory::new()->create(['user_id' => $user->id]);

// Option 2: You use the with method which will automaticall set the foreign key with the user id
$user = UserFactory::new()
    ->with(Recipe::class, 'recipes')
    ->create();

// Option 3: "make" the recipe first, this will NOT store it in the DB and therefore no problem with a foreign key not set, then you can "save" the recipe through the laravel relationship
$user = UserFactory::new()->create();
$user->recipes()->save(RecipeFactory::new()->make();

```

So to sum it up, I would NOT define it there. This way you are more flexible to which user this recipe belongs. Does this help?

**Comment by stephanus-tantiono** [comment]
Yes. Thank you very much for your help.

**Comment by devmsh** [comment]
@christophrumpel sorry for commenting in a close issue, but it seems the right place because you already discuss some options above

 ```php
$user = UserFactory::new()->create(); // you can use create because no foreign key on this model

// Option 1: Then you can create a new recipe and add the user id
RecipeFactory::new()->create(['user_id' => $user->id]);

// Option 2: You use the with method which will automaticall set the foreign key with the user id
$user = UserFactory::new()
    ->with(Recipe::class, 'recipes')
    ->create();

// Option 3: "make" the recipe first, this will NOT store it in the DB and therefore no problem with a foreign key not set, then you can "save" the recipe through the laravel relationship
$user = UserFactory::new()->create();
$user->recipes()->save(RecipeFactory::new()->make();
```

I think what @stephanus-stuff means and what I also miss when I tried to refactor my current factories to the new reloaded syntax is the default belongTo relation value. here is an example:

```php
$factory->define(Page::class, function (Faker $faker) {
    return [
        'story_id' => factory(Story::class),
        'chapter_id' => factory(Chapter::class),
        'content' => $faker->text,
        'page_number' => $faker->numberBetween(1, 20),
        'voice' => $faker->imageUrl(),
    ];
});
```

As you now, using this syntax, we ask Laravel to create a new story and chapter only and only if we did not provide an override value, this is the shorter new syntax for using function-based lazy loading technique.

In many tests is just call `factory(Page::class)->create()` and it will generate the corresponding story and chapter, and sometimes I already have the chapter or the story, so I pass them to the page factory.

The package is really amazing, but this issue prevents me from being able to refactor my tests without adding a complicated arrange phase, and this is completely inconsistent with the package goal if I got it right?
